inventory/packages/wua_windows.go

```diff
@@ -15,15 +15,139 @@ package packages
 )
 
-type (
-	IUpdateSession    = *ole.IDispatch
-	IUpdateCollection = *ole.IDispatch
-)
+var wuaSession sync.Mutex
+
+// IUpdateSession is a an IUpdateSession.
+type IUpdateSession struct {
+	obj *ole.IUnknown
+	ses *ole.IDispatch
+}
+
+func NewUpdateSession() (*IUpdateSession, error) {
+	wuaSession.Lock()
+	if err := ole.CoInitializeEx(0, ole.COINIT_MULTITHREADED); err != nil {
+		return nil, err
+	}
+
+	updateSessionObj, err := oleutil.CreateObject("Microsoft.Update.Session")
+	if err != nil {
+		return nil, fmt.Errorf(`oleutil.CreateObject("Microsoft.Update.Session"): %v`, err)
+	}
+
+	session, err := updateSessionObj.IDispatch(ole.IID_IDispatch)
+	if err != nil {
+		return nil, err
+	}
+	return &IUpdateSession{obj: updateSessionObj, ses: session}, nil
+}
+
+func (s *IUpdateSession) Close() {
+	wuaSession.Unlock()
+	ole.CoUninitialize()
+	s.ses.Release()
+	s.obj.Release()
+}
+
+// InstallWUAUpdate install a WIndows update.
+func (s *IUpdateSession) InstallWUAUpdate(updt *IUpdate) error {
+	title, err := updt.GetProperty("Title")
+	if err != nil {
+		return fmt.Errorf(`updt.GetProperty("Title"): %v`, err)
+	}
+
+	updateCollObj, err := oleutil.CreateObject("Microsoft.Update.UpdateColl")
+	if err != nil {
+		return fmt.Errorf(`oleutil.CreateObject("updateColl"): %v`, err)
+	}
+	defer updateCollObj.Release()
+
+	updateColl, err := updateCollObj.IDispatch(ole.IID_IDispatch)
+	if err != nil {
+		return err
+	}
+	defer updateColl.Release()
+	updts := &IUpdateCollection{updateColl}
+
+	eula, err := updt.GetProperty("EulaAccepted")
+	if err != nil {
+		return fmt.Errorf(`updt.GetProperty("EulaAccepted"): %v`, err)
+	}
+	// https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-oaut/7b39eb24-9d39-498a-bcd8-75c38e5823d0
+	if eula.Val == 0 {
+		DebugLogger.Printf("%s - Accepting EULA", title.Value())
+		if _, err := updt.CallMethod("AcceptEula"); err != nil {
+			return fmt.Errorf(`updateColl.CallMethod("AcceptEula"): %v`, err)
+		}
+	} else {
+		DebugLogger.Printf("%s - EulaAccepted: %v", title.Value(), eula.Value())
+	}
+
+	if _, err := updateColl.CallMethod("Add", updt); err != nil {
+		return fmt.Errorf(`updateColl.CallMethod("Add", updt): %v`, err)
+	}
+
+	DebugLogger.Printf("Downloading update %s", title.Value())
+	if err := s.DownloadWUAUpdateCollection(updts); err != nil {
+		return fmt.Errorf("DownloadWUAUpdateCollection error: %v", err)
+	}
+
+	DebugLogger.Printf("Installing update %s", title.Value())
+	if err := s.InstallWUAUpdateCollection(updts); err != nil {
+		return fmt.Errorf("InstallWUAUpdateCollection error: %v", err)
+	}
+
+	return nil
+}
+
+type IUpdateCollection struct {
+	*ole.IDispatch
+}
+
+type IUpdate struct {
+	*ole.IDispatch
+}
+
+func (c *IUpdateCollection) Add(updt *IUpdate) error {
+	if _, err := c.CallMethod("Add", updt); err != nil {
+		return fmt.Errorf(`updateColl.CallMethod("Add", updt): %v`, err)
+	}
+	return nil
+}
+
+func (c *IUpdateCollection) RemoveAt(i int) error {
+	if c == nil {
```

Comment:
here and below, is this an expected state? When would this happen and should it have comments for when/why it's needed?

Reply:
This is not expected and was here for my testing